core/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTextAggregator.java

```diff
@@ -0,0 +1,271 @@
+import org.elasticsearch.search.internal.ContextIndexSearcher;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Collections.emptyList;
+
+public class SignificantTextAggregator extends BucketsAggregator {
+    
+    private final StringFilter includeExclude;
+    protected final BucketCountThresholds bucketCountThresholds;
+    protected long numCollectedDocs;
+    private final BytesRefHash bucketOrds;
+    private final SignificanceHeuristic significanceHeuristic;
+    private SignificantTextAggregatorFactory termsAggFactory;
+    private final DocValueFormat format = DocValueFormat.RAW;
+    private final String fieldName;
+    private DuplicateByteSequenceSpotter dupSequenceSpotter ;
+    private long lastTrieSize;
+
+    private final boolean filterDuplicateText;
+    private static final int MEMORY_GROWTH_REPORTING_INTERVAL_BYTES = 5000;
+
+
+
+    public SignificantTextAggregator(String name, AggregatorFactories factories,
+            SearchContext context, Aggregator parent, List<PipelineAggregator> pipelineAggregators,
+            BucketCountThresholds bucketCountThresholds, IncludeExclude.StringFilter includeExclude,
+            SignificanceHeuristic significanceHeuristic, SignificantTextAggregatorFactory termsAggFactory,
+            String fieldName, boolean filterDuplicateText,
+            Map<String, Object> metaData) throws IOException {
+        super(name, factories, context, parent, pipelineAggregators, metaData);
+        this.bucketCountThresholds = bucketCountThresholds;
+        this.includeExclude = includeExclude;
+        this.filterDuplicateText = filterDuplicateText;
+        this.significanceHeuristic = significanceHeuristic;
+        this.termsAggFactory = termsAggFactory;
+        this.fieldName = fieldName;
+        bucketOrds = new BytesRefHash(1, context.bigArrays());
+        dupSequenceSpotter = new DuplicateByteSequenceSpotter();        
+        lastTrieSize = dupSequenceSpotter.getEstimatedSizeInBytes();
+    }
+
+    
+    
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+            final LeafBucketCollector sub) throws IOException {
+        final ShardGetService shardGetService = super.context.indexShard().getService();
+        final BytesRefBuilder previous = new BytesRefBuilder();
+        return new LeafBucketCollectorBase(sub, null) {
+            
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                Terms tv = ctx.reader().getTermVector(doc, fieldName);
+                if (tv!=null &&  (tv.hasPositions()  || tv.hasOffsets())){
+                    // Not necessarily faster than re-analysis in my tests!
+                    processTokenStream(doc, bucket, new TokenStreamFromTermVector(tv, -1), "");
```

Comment:
Given their questionable performance, yes it would make sense to have an opt-in parameter in the query request. I can remove if we're just not convinced.